doc/notes/request.md

```diff
@@ -0,0 +1,57 @@
+## MPIR_Request Objects
+
+The request object is the most critical object in the MPICH communication
+path. All our MPID interface are non-blocking, thus they all follow the
+pattern of --
+1. creating request objects
+1. setup `cc_ptr`, ref_count, store necessary information
+1. issue operation using the request as context
+1. [progress]
+1. complete or update the request as progress are made and relevant callbacks
+or event function gets called
+1. MPI_Wait or MPI_Test return success and reaps the request object
+
+The request struct need hold necessary fields for all kinds of communications,
+and all layers need touch the request object throughout its lifetime. Thus,
+the request struct consists fields complexed by both the communication types
+and code layers. It is by far the most complex part of MPICH code -- thus the
+need for this separate note.
+
+The MPIR_Request struct is defined in
+[mpir_request.h](../src/include/mpir_request.h#L168-L225). Other than the
```

Comment:
Link does not work. It should be
`../../src/include/mpir_request.h#L168-L225`

Reply:
I moved it to `doc/wiki/notes` and fixed up the link